File: Scripts/Script_lines.py
import json


def preprocess_json(input_json_file, output_json_file):
    try:
        with open(input_json_file, 'r', encoding='utf-8') as file:
            original_data = json.load(file)
    except Exception as e:
        print(f"Error reading or loading the JSON file: {e}")
        exit(1)

    transformed_data = []
    for line_name, details in original_data.items():
        new_entry = {
            "Chinese_name": line_name
        }
        new_entry.update(details)
        transformed_data.append(new_entry)

    try:
        with open(output_json_file, 'w', encoding='utf-8') as file:
            json.dump(transformed_data, file, ensure_ascii=False, indent=4)
    except Exception as e:
        print(f"Error writing or saving the transformed JSON file: {e}")
        exit(1)


def generate_sql(input_json_file, output_sql_file):
    create_table = """
    create table if not exists lines
    (
        line_id serial  primary key,
        Chinese_name varchar(50) not null unique,
        start_time varchar(50) not null,
        end_time varchar(50) not null,
        mileage numeric(10,2) not null,
        color varchar(20) not null,
        first_opening varchar(50) not null,
        intro varchar(5000) not null,
        url varchar(100) not null
    );
    """

    try:
        with open(input_json_file, 'r', encoding='utf-8') as file:
            data = json.load(file)
    except Exception as e:
        print(f"Error reading or loading the JSON file: {e}")
        exit(1)

    try:
        with open(output_sql_file, 'w', encoding='utf-8') as file:
            file.write(create_table + "\n")
            for record in data:
                chinese_name = record["Chinese_name"].replace("'", "''")
                start_time = record["start_time"].replace("'", "''")
                end_time = record["end_time"].replace("'", "''")
                mileage = record["mileage"]
                color = record["color"].replace("'", "''")
                first_opening = record["first_opening"].replace("'", "''")
                intro = record["intro"].replace("'", "''")
                url = record["url"].replace("'", "''")

                insert_sql = (f"INSERT INTO lines "
                              f"(Chinese_name, start_time, end_time, mileage, color, first_opening, intro,  url) "
                              f"VALUES ('{chinese_name}', '{start_time}', '{end_time}',  {mileage}, '{color}',"
                              f" '{first_opening}', '{intro}', '{url}');\n")
                file.write(insert_sql)
    except Exception as e:
        print(f"Error writing or saving the transformed JSON file: {e}")


json_input_path = '../resource/lines.json'  # original JSON file path
json_output_path = '../out/lines_adjusted.json'  # adjusted JSON file path
sql_output_path = '../out/lines.sql'  # SQL file path

preprocess_json(json_input_path, json_output_path)
generate_sql(json_output_path, sql_output_path)

File: Scripts_in_Java/Script_lines.java
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.google.gson.stream.JsonWriter;

import java.io.*;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class Script_lines {

    public static void preprocessJson(String inputJsonFile, String outputJsonFile) {
        try (Reader reader = new FileReader(inputJsonFile)) {
            Type type = new TypeToken<Map<String, Map<String, Object>>>(){}.getType();
            Map<String, Map<String, Object>> originalData = new Gson().fromJson(reader, type);
            List<Map<String, Object>> transformedData = new ArrayList<>();

            originalData.forEach((lineName, details) -> {
                details.put("Chinese_name", lineName);
                transformedData.add(details);
            });

            try (Writer writer = new FileWriter(outputJsonFile);
                 JsonWriter jsonWriter = new JsonWriter(writer)) {
                jsonWriter.setIndent("    ");
                new Gson().toJson(transformedData, List.class, jsonWriter);
            }

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static void generateSql(String inputJsonFile, String outputSqlFile) {
        String createTable = """
            create table if not exists lines
            (
                line_id serial primary key,
                Chinese_name varchar(50) not null unique,
                start_time varchar(50) not null,
                end_time varchar(50) not null,
                mileage numeric(10,2) not null,
                color varchar(20) not null,
                first_opening varchar(50) not null,
                intro varchar(5000) not null,
                url varchar(100) not null
            );
            """;

        try (BufferedWriter writer = new BufferedWriter(new FileWriter(outputSqlFile));
             Reader reader = new FileReader(inputJsonFile)) {
            writer.write(createTable + "\n");
            Type listType = new TypeToken<List<Map<String, Object>>>(){}.getType();
            List<Map<String, Object>> data = new Gson().fromJson(reader, listType);

            data.forEach(record -> {
                try {
                    String Chinese_name = ((String) record.get("Chinese_name")).replace("'", "''");
                    String start_time = ((String) record.get("start_time")).replace("'", "''");
                    String end_time = ((String) record.get("end_time")).replace("'", "''");
                    String temp = (String) record.get("mileage");
                    double mileage = Double.parseDouble(temp);
                    String color = ((String) record.get("color")).replace("'", "''");
                    String first_opening = ((String) record.get("first_opening")).replace("'", "''");
                    String intro = ((String) record.get("intro")).replace("'", "''");
                    String url = ((String) record.get("url")).replace("'", "''");

                    String insertSql = String.format("INSERT INTO lines (Chinese_name, start_time, end_time, mileage, color, first_opening, intro, url) VALUES ('%s', '%s', '%s', %.2f, '%s', '%s', '%s', '%s');\n",
                            Chinese_name, start_time, end_time, mileage, color, first_opening, intro, url);
                    writer.write(insertSql);
                } catch (IOException e) {
                    System.out.println("Error reading or writing JSON file: " + e.getMessage());
                }
            });
        } catch (IOException e) {
            System.out.println("Error reading or writing JSON file: " + e.getMessage());
        }
    }

    public static void main(String[] args) {
        String jsonInputPath = "C:\\Users\\Eric\\IdeaProjects\\CS307\\Project1\\resource\\lines.json";
        String jsonOutputPath = "C:\\Users\\Eric\\IdeaProjects\\CS307\\Project1\\out\\lines_adjusted.json";
        String sqlOutputPath = "C:\\Users\\Eric\\IdeaProjects\\CS307\\Project1\\out\\lines.sql";

        preprocessJson(jsonInputPath, jsonOutputPath);
        generateSql(jsonOutputPath, sqlOutputPath);
    }
}
